### ISSUE TITLE
Factory methods for State

### ISSUE THREAD

**Issue by martinvw** [eclipse/smarthome#4612]
I noticed that quite some bindings have there own utility class / methods to do a null check and conditionally convert it to either ~~`UnDefType.NULL`~~ _(edit: `UnDefType.UNDEF`)_ or the respective value, some examples:

[netatmo ChannelTypeUtils](https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/ChannelTypeUtils.java)
[NestBaseHandler @ line 109](https://github.com/openhab/openhab2-addons/blob/8f7c32ce13b89e0bd913758e158b5b55eae7ecd6/addons/binding/org.openhab.binding.nest/src/main/java/org/openhab/binding/nest/handler/NestBaseHandler.java#L109)
[KodiHandler.java @ Line 399](https://github.com/openhab/openhab2-addons/blob/f238e11e8b61def4d1247e37df32be45a319377f/addons/binding/org.openhab.binding.kodi/src/main/java/org/openhab/binding/kodi/handler/KodiHandler.java#L399)

I think that this should/could be the responsibility of the framework, imho every method (for which it is sensible) should have one or more nullable `valueOf` methods. Which return either ~~`UnDefType.NULL`~~ _(edit: `UnDefType.UNDEF`)_ or the wrapper class of the requested type.

I would also suggest to change the meaning of [DecimalType.valueOf](https://github.com/eclipse/smarthome/blob/36c9e2717e4f89cf60a0d885e7ecc6069b338f48/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/DecimalType.java#L68) to allow null and handle it properly.

I can / would like to make this PR but if people already reject it would be waste of time, so what do you think?

CC: @maggu2810 @sjka @htreu 

PS: maybe I would even like to mark the constructors deprecated for touched classes...

**Comment by maggu2810** [comment]
IIRC `UnDefType.NULL` should not be set by the binding itself.

* [Kodi State createState(String string)]: I don't think that an empty string should be converted to `UnDefType.NULL` in general. A empty string wrapped by a string type is AFAIK valid and IMHO there is a difference between an empty string and null.
* Isn't the absence of a "internal value" (internal by the binding and its connection to the outside handling) and its interpretation specific to the binding itself?
* Why is the reference `null` at all? If the communication to the real device broken, not yet complete? If a state if not known shouldn't it be set to `UnDefType.UNDEF` by the binding?
* Isn't it in the responsible of the specific binding to decide if no new value is interpret as missing or it would like to keep the old value? Okay this could be still applied...

[Kodi State createState(String string)]: https://github.com/openhab/openhab2-addons/blob/f238e11e8b61def4d1247e37df32be45a319377f/addons/binding/org.openhab.binding.kodi/src/main/java/org/openhab/binding/kodi/handler/KodiHandler.java#L399

**Comment by martinvw** [comment]
Hello @maggu2810 ,

I do agree with your first point about empty strings. 

I was not aware of the fact that `UnDefType.NULL` should not be used in the bindings, does the same hold for the other `UnDefType` values?

A lot of bindings are relatively thin (which is a good thing) but if they just want to be a proxy thing between openHAB and a remote API.  Then everyone of them needs not only to convert their string / boolean or number to Smarthome types but also has to do that conditionally. 

Because you are not allowed to pass `null` in a `DecimalType`, `DateTimeType `and `StringListType`. The StringType does not really care at first but will throw an NPE is someone calls the `hashcode `method.

I think that in the serial world the `null `might really mean that something went wrong in the communication but in REST world, it is just the provider telling you that it does not have a value (we cannot know why) and because that has broken a lot bindings already those bindings now all contain the same logic.

>Isn't it in the responsible of the specific binding to decide if no new value is interpret as missing or it would like to keep the old value? Okay this could be still applied...

So that could result in two tastes:

```java
Result result = gson.fromString(result, Result.class);

state = result.temperature == null ? UnDefType.UNDEF : new DecimalType(result.temperature);
updateState(TEMPERATURE, state);

// or
if ( result.temperature != null) {
   updateState(TEMPERATURE, new DecimalType(result.temperature));
}
```

So can we reach a conclusion about the following questions:

1. What `UnDefType` values can be used by bindings 
2. Do we prefer one of both options above or that they both have their place? 
3 . And if the first one has its place, can we put that burden into eclipse/smarthome?

**Comment by maggu2810** [comment]
AFAIK:
* A state must be not `null`.
* The state is initialized by the framework with `UnDefType.NULL`. This enum value is only used by the framework for an initial value.
* `UnDefType.UNDEF` could be used by a binding to signal that a state is not defined.

Isn't it then dependent of the proxy implementation if a missing state would like to be interpret as "keep the old value" or I don't know the value anymore?

**Comment by martinvw** [comment]
> Isn't it then dependent of the proxy implementation if a missing state would like to be interpret as "keep the old value" or I don't know the value anymore?

Yes it is, so I think both tastes above have their place and the `null`-aware factory method could still have value but should return `UNDEF `instead of `NULL`

> The state is initialized by the framework with `UnDefType.NULL`. This enum value is only used by the framework for an initial value.
> `UnDefType.UNDEF` could be used by a binding to signal that a state is not defined.

So we should do an cleanup in openHAB and maybe smarthome? Will the `UNDEF  `behave the same inside rules or is this a breaking change?

**Comment by kaikreuzer** [comment]
In general I agree with @maggu2810's summary, I have just one comment regarding

>  This enum value [NULL] is only used by the framework for an initial value.

This is correct in practice right now and it is definitely the main use case for it. But according to the JavaDoc, the definition is:

```
 * There are situations when item states do not have any defined value.
 * This might be because they have not been initialized yet (never
 * received an state update so far) or because their state is ambiguous
 * (e.g. a dimmed light that is treated as a switch (ON/OFF) will have
 * an undefined state if it is dimmed to 50%).
```
which means:
- NULL = no state
- UNDEF = ambiguous, i.e. not properly determinable state

While for regular use cases, there always is some state (a light is on or off or whatever, but it always has a state), there are a few channels where one could argue that they do not have ANY state at certain times, e.g. the "current artist" if a player is stopped. I therefore would say that we could allow bindings to set NULL for such rare use cases. Wdyt?

> Will the UNDEF behave the same inside rules or is this a breaking change?

No, if rules currently check a state against UnDefType.NULL, this will break. But I doubt that this is a common use case.

**Comment by martinvw** [comment]
> While for regular use cases, there always is some state (a light is on or off or whatever, but it always has a state), there are a few channels where one could argue that they do not have ANY state at certain times, e.g. the "current artist" if a player is stopped. I therefore would say that we could allow bindings to set NULL for such rare use cases. Wdyt?

So the AllPlayHandler.java would still set `NULL` sounds reasonable.

@kaikreuzer what do you think about my earlier questions:

2. Should in case of `null` no update to the state be requested, or an `UnDefType` or do both options have there own use cases
3. Do you agree that for a use case the (some) `UnDefType` or some concrete `State` will make sense?

**Comment by martinvw** [comment]
@maggu2810 @kaikreuzer @sjka any updates / opinions about this?

**Comment by maggu2810** [comment]
> I therefore would say that we could allow bindings to set NULL for such rare use cases. Wdyt?

```
 * There are situations when item states do not have any defined value.
 * This might be because they have not been initialized yet (never
 * received an state update so far) or because their state is ambiguous
 * (e.g. a dimmed light that is treated as a switch (ON/OFF) will have
 * an undefined state if it is dimmed to 50%).
```

WDYT about:

* NULL = never received an state update, so only used by the framework itself
* UNDEF = all other reasons and used by bindings (e.g. ambiguous, undefined state, ...)

So, if there is no "current artist" we could also set undefined, because there there is no one and so the artist is not defined.

**Comment by martinvw** [comment]
For me it feels good to have a separation between what he framework sets and what bindings could /should set. So 👍On my side

**Comment by sjka** [comment]
👍  from my side too. 

Your suggestion matches my current understanding - but admittedly that's mainly because I hardly ever read Javadoc 😉 

I didn't spent as much thought on this topic as Kai did, so I'd appreciate @kaikreuzer's verdict here.

**Comment by kaikreuzer** [comment]
As mentioned above: The original idea was to say that UNDEF stands for "there is a state, but I cannot express it" - e.g. because you ask to retrieve a 50% state as an OpenClosedType.
NULL was meant to have the very same meaning as "null" in Java, i.e. "there is nothing / no state at all".

> So, if there is no "current artist" we could also set undefined

... while NULL would be the better match.

> it feels good to have a separation between what he framework sets and what bindings could /should set.

If you can replace this "feeling" by some objective arguments why this separation is a good thing, I might be tempted to give in. But for the moment, I am not yet convinced, sorry ;-)

**Comment by martinvw** [comment]
No problem, given this is not my original question, that was whether it would make sense to have a factory method in smarthome that return some either some undefined value or the non null value of the requested type. So what do you think about that one 😉

**Comment by kaikreuzer** [comment]
You are right, we should discuss your original question.
I think the discussion already highlighted that there wouldn't really be a one-fits-all implementation of such methods as the right behavior can depend on the concrete binding.
Imho there are these three potential options of which each channel picks one:
Upon "null" from the remote device
- don't send any state update 
- send UnDefType.UNDEF
- send UnDefType.NULL

> imho every method (for which it is sensible) should have one or more nullable valueOf methods

Note that EVERY Type class already has a `valueOf(String)` method as this is required by the `TypeParser`. You are right that this is currently probably not null safe, so the implementations should be improved/fixed. But I think the correct return value would simply be `null` in such a case.

**Comment by martinvw** [comment]
> Note that EVERY Type class already has a valueOf(String) method as this is required by the TypeParser. You are right that this is currently probably not null safe, so the implementations should be improved/fixed. But I think the correct return value would simply be null in such a case.

If you don't want send updates if the input to the `valueOf` is `null` it's easy to check that. But if you want send the updates as is, you will always need a util like class like we have a lot already or a lot repeated if checks.

Imho the `valueOf` returns an 'incorrect' value because because 'all' (or most) the methods that accept State don't want `null` but want to rather receive an `UnDefType` in such cases. 

But if you do not agree with this, could we then maybe meet in the middle by adding overloaded `valueOf` (or maybe `valueOfOr`) methods with a default / fallback parameter. That could handle the scenarios 'send UnDefType.UNDEF', 'send UnDefType.NULL'. So we do not need all the `null` handling in all bindings, which just want to fallback to another value.

**Comment by kaikreuzer** [comment]
Let me rephrase my answer: From the 3 options above, I would claim that in 95% of the cases, the correct behavior is (1), i.e. immediately stop any further processing. "'all' (or most) the methods that accept State" should therefore not be called at all and where ever we find a binding to create an UnDefType.NULL state and pass this on, we should really question whether this is the right thing to do there.
I would therefore claim that such "helper methods" would rather help hiding bugs.

**Comment by martinvw** [comment]
Please note that some of your statement above have a pretty big impact, if the `valueOf` methods should be `null` safe and return `null` their output cannot be passed on
`updateState(CHANNEL_BB, OnOffType.valueOf(response.getBB());` without suppressing any warnings. Because that method does not accept `null`.

Imho this will lead to different ugly code because it will move the same problem one layer up ;-) 

Because we state there is no safe way to just 'proxy' a json api to smarthome:

I would prefer something like this for simple cases:

```java
updateState(CHANNEL_AA, OnOffType.valueOf(response.getAA());
updateState(CHANNEL_BB, OnOffType.valueOf(response.getBB());
updateState(CHANNEL_CC, OnOffType.valueOf(response.getCC());
```

But it will look more like

```java
if (response.getAA() != null) {
   // maybe even like this :'-(
   @NotNull
   State value = OnOffType.valueOf(response.getAA());
   updateState(CHANNEL_AA, value);
}
if (response.getBB() != null) {
    updateState(CHANNEL_BB, OnOffType.valueOf(response.getBB());
}
if (response.getCC() != null) { 
    updateState(CHANNEL_CC, OnOffType.valueOf(response.getCC());
}
```

Which will lead me to a new desire of having an `updateStateOrIgnore` (for the sake of DRY) method which either does an `updateState` or ignores it if the value is null this method should of course have a nullable `state` as the second parameter.

**Comment by kaikreuzer** [comment]
> their output cannot be passed on `updateState(CHANNEL_BB, OnOffType.valueOf(response.getBB())`

Perfect, just what I want to achieve 😎 

> because it will move the same problem one layer up

No, it will make the implementor think about what he is doing!

```
if (response.getAA() != null) {
   // maybe even like this :'-(
   @NotNull
   State value = OnOffType.valueOf(response.getAA());
   updateState(CHANNEL_AA, value);
}
```
This is actually not correct. The better version would be:
```
String receivedValue = response.getAA();
if (receivedValue != null) {
   updateState(CHANNEL_AA, OnOffType.valueOf(receivedValue));
}
```

> Which will lead me to a new desire of having an updateStateOrIgnore

Well, this is a very different request to the orginal one now.
And I am not sure if such a method would be really that helpful. After all, null checks are a very common thing in Java and trying to avoid them by introducing additional methods that do that check might not always be the best thing. If a binding has 20 channels and the code indeed repeats itself, it is imho ok to have a helper method within this binding (which can then also be specific for the Java types of the received values, which can also be something different for different bindings).

**Comment by martinvw** [comment]
@kaikreuzer 

> This is actually not correct. The better version would be:

Nope, because it will fail because of `null` checks.

**Comment by martinvw** [comment]
> Well, this is a very different request to the orginal one now.
And I am not sure if such a method would be really that helpful.

Let's agree to disagree then about my main request...

If we decide that the `valueOf` methods should be `null` safe and return `null` I can create a PR with some tests for that.

**Comment by kaikreuzer** [comment]
> Nope, because it will fail because of null checks.

At which line?

> If we decide that the valueOf methods should be null safe and return null I can create a PR with some tests for that.

Well, considering the example code above, we would then probably rather want to add a `@NonNull` annotation to its parameter...?

**Comment by martinvw** [comment]
>> Nope, because it will fail because of null checks.

>At which line?

Line 3 because it pass a Nullable value to a NonNull parameter.

**Comment by kaikreuzer** [comment]
Well, not if we add `@NonNull` annotations to the valueOf method (not only its parameter, but also its return value).

**Comment by martinvw** [comment]
> Well, not if we add @NonNull annotations to the valueOf method (not only its parameter, but also its return value).

That will be pretty big and breaking, should we add a follow issue to track that?

**Comment by kaikreuzer** [comment]
Are you sure that it is breaking? Wouldn't it just document the current behavior?
As I said, you currently must not pass a null value to it and it also never returns null, but rather throws an IAE.

**Comment by kaikreuzer** [comment]
It would probably even suffice to only add `@NonNull` on the return value and simply throw IAEs if null is passed in as a parameter.

**Comment by kaikreuzer** [comment]
@martinvw This was a long discussion, but did we actually have any outcome? Can we close it as is or do you still think that something should be addressed here?

**Comment by martinvw** [comment]
I’m okay closing this
